oml/metrics/embeddings.py

```diff
@@ -166,47 +156,42 @@ def compute_metrics(self) -> TMetricsDict_ByLabels:  # type: ignore
 
         self._compute_retrieval_results()
 
-        metrics: TMetricsDict_ByLabels = dict()
+        args_r = {
+            "cmc_top_k": self.cmc_top_k,
+            "precision_top_k": self.precision_top_k,
+            "map_top_k": self.map_top_k,
+            "rr": self.retrieval_results,
+            "reduce": False,
+        }
 
-        # note, here we do micro averaging
-        metrics[self.overall_categories_key] = calc_retrieval_metrics_rr(
-            rr=self.retrieval_results,
-            cmc_top_k=self.cmc_top_k,
-            precision_top_k=self.precision_top_k,
-            map_top_k=self.map_top_k,
-            reduce=False,
-        )
-
-        embeddings = self.acc.storage[self._acc_embeddings_key]
-        metrics[self.overall_categories_key].update(calc_topological_metrics(embeddings, self.pcf_variance))
+        args_t = {"embeddings": self.acc.storage[self._acc_embeddings_key], "pcf_variance": self.pcf_variance}
 
         if CATEGORIES_COLUMN in self.dataset.extra_data:
             categories = np.array(self.dataset.extra_data[CATEGORIES_COLUMN])
-            ids_query = self.dataset.get_query_ids()
-            query_categories = categories[ids_query]
+            query_categories = categories[self.dataset.get_query_ids()]
 
-            for category in np.unique(query_categories):
-                mask_query_sz = query_categories == category
-                metrics[category] = take_unreduced_metrics_by_mask(metrics[self.overall_categories_key], mask_query_sz)
+            metrics_r = calc_retrieval_metrics_rr(query_categories=query_categories, **args_r)  # type: ignore
+            metrics_t = calc_topological_metrics(categories=categories, **args_t)  # type: ignore
 
-                mask_dataset_sz = categories == category
-                metrics[category].update(calc_topological_metrics(embeddings[mask_dataset_sz], self.pcf_variance))
+            self.metrics_unreduced = {cat: {**metrics_r[cat], **metrics_t[cat]} for cat in metrics_r.keys()}
```

Comment:
On this line `metrics_r` have all retrieval metrics and categories as top-level keys because of [this](https://github.com/OML-Team/open-metric-learning/blob/521df23cb0e9231a48f0251da95106a413e5afac/oml/functional/metrics.py#L67), so
1) You can't do `metrics_t[cat]` for retrieval metric keys
2) `self.metrics_unreduced` won't include `metrics_t`

Reply:
discussed offline